lib/AST/ASTImporter.cpp

```diff
@@ -52,9 +52,57 @@ namespace clang {
                           public StmtVisitor<ASTNodeImporter, Stmt *> {
     ASTImporter &Importer;
 
+    // Wrapper for an overload set.
+    template <typename ToDeclT> struct CallOverloadedCreateFun {
+      template <typename... Args>
+      auto operator()(Args &&... args)
+          -> decltype(ToDeclT::Create(std::forward<Args>(args)...)) {
+        return ToDeclT::Create(std::forward<Args>(args)...);
+      }
+    };
+
+    // Always use this function to create a Decl during import. There are
+    // certain tasks which must be done after the Decl was created, e.g. we
+    // must immediately register that as an imported Decl.
+    template <typename ToDeclT, typename FromDeclT, typename... Args>
+    std::pair<ToDeclT *, bool> CreateDecl(FromDeclT *FromD, Args &&... args) {
+      // There may be several overloads of ToDeclT::Create. We must make sure
+      // to call the one which would be chosen by the arguments, thus we use a
+      // wrapper for the overload set.
+      CallOverloadedCreateFun<ToDeclT> OC;
```

Comment:
I am not sure we need a wrapper. A lambda or casting to a function pointer might do as well.

Reply:
Yeah, my first attempt was exactly to create a function pointer. I there were no overloads  XXXDecl::Create, that would work. However there are Decl subclasses where `::Create` is overloaded. We can't have a pointer to an overloaded function without knowing its signature. we can't have a pointer to an overload set. This solution delays the overload resolution up to the point of the call.
References: https://stackoverflow.com/questions/9054703/overloaded-function-as-argument-of-variadic-template-function